lib/portlayer/metrics/container.go

```diff
@@ -66,113 +62,117 @@ type MemoryMetrics struct {
 	Interval int32
 }
 
-// collectorVM is the VM metrics collector
-type collectorVM struct {
+// vmCollector is the VM metrics collector
+type vmCollector struct {
 	perfMgr *performance.Manager
 
+	timer     *time.Ticker
+	startTime time.Time
+	stopper   chan bool
+
 	// subscribers to streaming
 	mu   sync.RWMutex
-	subs map[types.ManagedObjectReference]*subscriber
-}
-
-// subscriber is the receiver of the metrics
-type subscriber struct {
-	id        string
-	ref       types.ManagedObjectReference
-	publisher *pubsub.Publisher
+	subs map[interface{}]*subscription
 }
 
-func newVMCollector(session *session.Session) *collectorVM {
-	s := make(map[types.ManagedObjectReference]*subscriber)
-	cc := &collectorVM{
-		subs:    s,
+// newVMCollector will instantiate a new collector responsible for
+// gathering VM metrics
+func newVMCollector(session *session.Session) *vmCollector {
+	cc := &vmCollector{
+		subs:    make(map[interface{}]*subscription),
 		perfMgr: performance.NewManager(session.Vim25()),
+		stopper: make(chan bool),
 	}
-
-	// kick off the subscription sampler
-	go cc.sampler()
-
 	return cc
 }
 
-// sampler will check for subscriptions and sample when needed
-func (cc *collectorVM) sampler() {
-	// TODO: replace this with service that only runs when there are
-	// active subscribers
-	d := time.Duration(int64(sampleInterval)) * time.Second
-	for range time.Tick(d) {
-		// if we have no subscribers skip
-		if cc.SubscriberCount() == 0 {
-			continue
+// Start will begin the collection polling process
+func (cc *vmCollector) Start() {
+	defer trace.End(trace.Begin(""))
+	// create timer with sampleInterval alignment
+	cc.timer = time.NewTicker(time.Duration(int64(sampleInterval)) * time.Second)
+	// loop on the channel
+	go func() {
+		cc.startTime = time.Now()
+		// initial collection and then at specified interval
+		cc.collect(cc.Subscribers())
+		for {
+			select {
+			case <-cc.timer.C:
+				// collect metrics for current subscribers
+				cc.collect(cc.Subscribers())
+			case <-cc.stopper:
+				// Ticker has been stopped, exit routine
+				return
+			}
 		}
-		// collect metrics for current subscribers
-		cc.collect(cc.Subscribers(), false)
-	}
+	}()
 }
 
-// Sample returns a single metrics collection
-func (cc *collectorVM) Sample(contain interface{}) (chan interface{}, error) {
+// Stop will stop the collection polling process
+func (cc *vmCollector) Stop() {
 	defer trace.End(trace.Begin(""))
-
-	containerSub, err := containerSubscriber(contain)
-	if err != nil {
-		return nil, err
-	}
-
-	// create a publisher and subscribe
-	containerSub.publisher = pubsub.NewPublisher(100*time.Millisecond, 0)
-	ch := containerSub.publisher.Subscribe()
-
-	// create a map with this single subscriber
-	single := make(map[types.ManagedObjectReference]*subscriber)
-	single[containerSub.ref] = containerSub
-
-	// collect the metrics
-	go cc.collect(single, true)
-
-	return ch, nil
+	cc.timer.Stop()
+	cc.stopper <- true
```

Comment:
You can just close the channel here, won't be used after this anyhow.  The golang nerds would also tell you to use `chan struct{}` instead of `chan bool`, since it uses less memory.